pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java

```diff
@@ -117,43 +105,40 @@ public String toExplainString() {
 
   @Override
   protected TransferableBlock getNextBlock() {
-    BaseDataBlock dataTable;
+    BaseDataBlock dataBlock;
     TransferableBlock transferableBlock = null;
     boolean isEndOfStream;
-    if (_dataTableBlockBaseOperator != null) {
-      transferableBlock = _dataTableBlockBaseOperator.nextBlock();
-      dataTable = transferableBlock.getDataBlock();
-      isEndOfStream = TransferableBlockUtils.isEndOfStream(transferableBlock);
-    } else {
-      dataTable = _dataTable;
-      isEndOfStream = true;
-    }
+    transferableBlock = _dataTableBlockBaseOperator.nextBlock();
+    dataBlock = transferableBlock.getDataBlock();
+    isEndOfStream = TransferableBlockUtils.isEndOfStream(transferableBlock);
 
     try {
       switch (_exchangeType) {
-        // TODO: random and singleton distribution should've been selected in planning phase.
         case SINGLETON:
+          // TODO: singleton or random distribution should've been distinguished in planning phase.
```

Comment:
qq on single vs random --

Assuming `stage 0` is the root stage, the intermediary `stage 1` should be having `senders` sending to a single `receiver` in the root stage at all times because at the end there has to be a single receiver (ordered or unordered etc). If so, then I am not sure why we are currently doing random ?

Reply:
echo on https://github.com/apache/pinot/pull/9064#discussion_r928123071